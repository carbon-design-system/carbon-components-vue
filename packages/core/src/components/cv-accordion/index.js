import CvAccordion from './cv-accordion';
import CvAccordionItem from './cv-accordion-item';
import CvAccordionSkeleton from './cv-accordion-skeleton';

export { CvAccordion, CvAccordionItem, CvAccordionSkeleton };
