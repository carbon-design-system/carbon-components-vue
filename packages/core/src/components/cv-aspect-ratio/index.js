import CvAspectRatio from './cv-aspect-ratio';
export { CvAspectRatio };
