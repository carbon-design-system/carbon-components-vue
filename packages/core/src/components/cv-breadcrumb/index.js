import CvBreadcrumb from './cv-breadcrumb';
import CvBreadcrumbItem from './cv-breadcrumb-item';
import CvBreadcrumbSkeleton from './cv-breadcrumb-skeleton';

export { CvBreadcrumb, CvBreadcrumbItem, CvBreadcrumbSkeleton };
