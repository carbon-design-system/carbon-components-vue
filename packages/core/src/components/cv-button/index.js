import CvButton from './cv-button';
import CvIconButton from './cv-icon-button';
import CvButtonSkeleton from './cv-button-skeleton';
import CvButtonSet from './cv-button-set';

export { CvButton, CvIconButton, CvButtonSkeleton, CvButtonSet };
