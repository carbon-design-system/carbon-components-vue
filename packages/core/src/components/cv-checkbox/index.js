import CvCheckbox from './cv-checkbox';
import CvCheckboxSkeleton from './cv-checkbox-skeleton';
export { CvCheckbox, CvCheckboxSkeleton };
