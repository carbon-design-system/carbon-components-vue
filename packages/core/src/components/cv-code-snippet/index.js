import CvCodeSnippet from './cv-code-snippet';
import CvCodeSnippetSkeleton from './cv-code-snippet-skeleton';
export { CvCodeSnippet, CvCodeSnippetSkeleton };
