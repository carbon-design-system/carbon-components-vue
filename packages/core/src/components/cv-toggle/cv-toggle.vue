<template>
  <div :class="{ [`${carbonPrefix}--form-item`]: formItem }">
    <input
      v-bind="$attrs"
      v-on="inputListeners"
      :class="[`${carbonPrefix}--toggle-input`, { 'bx--toggle-input--small': small }]"
      type="checkbox"
      :id="uid"
      :checked="isChecked === true"
      :aria-checked="`${isChecked}`"
      :value="value"
      ref="input"
    />
    <label :class="`${carbonPrefix}--toggle-input__label`" :for="uid" :aria-label="hiddenLabel">
      {{ visibleLabel }}
      <span :class="`${carbonPrefix}--toggle__switch`">
        <svg :class="`${carbonPrefix}--toggle__check`" width="6px" height="5px" viewBox="0 0 6 5">
          <path d="M2.2 2.7L5 0 6 1 2.2 5 0 2.7 1 1.5z" />
        </svg>
        <span :class="`${carbonPrefix}--toggle__text--off`" aria-hidden="true">
          <slot name="text-left">Off</slot>
        </span>
        <span :class="`${carbonPrefix}--toggle__text--on`" aria-hidden="true">
          <slot name="text-right">On</slot>
        </span>
      </span>
    </label>
  </div>
</template>

<script>
import { uidMixin, checkMixin, carbonPrefixMixin, methodsMixin } from '../../mixins';

export default {
  name: 'CvToggle',
  mixins: [uidMixin, checkMixin, carbonPrefixMixin, methodsMixin({ input: ['blur', 'focus'] })],
  inheritAttrs: false,
  props: {
    small: Boolean,
    label: String,
    formItem: { type: Boolean, default: true },
    hideLabel: Boolean,
  },
  computed: {
    hiddenLabel() {
      return this.hideLabel ? this.label : undefined;
    },
    visibleLabel() {
      return this.hideLabel ? undefined : this.label;
    },
  },
};
</script>
