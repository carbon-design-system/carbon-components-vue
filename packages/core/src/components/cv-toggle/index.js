import CvToggle from './cv-toggle';
import CvToggleSkeleton from './cv-toggle-skeleton';
export { CvToggle, CvToggleSkeleton };
