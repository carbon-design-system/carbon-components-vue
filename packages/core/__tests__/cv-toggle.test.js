// import { shallowMount as shallow } from '@vue/test-utils';
import { testComponent, awaitNextTick } from './_helpers';
const { shallowMount: shallow, trigger, setProps } = awaitNextTick;
import { CvToggle } from '@/components/cv-toggle';
// import { settings as carbonSettings } from 'carbon-components';

// const carbonPrefix = carbonSettings.prefix;

describe('CvToggle', () => {
  // ***************
  // PROP CHECKS
  // ***************
  testComponent.propsAreType(CvToggle, ['label', 'name', 'value'], String);
  testComponent.propsAreType(CvToggle, ['formItem', 'checked', 'small', 'hideLabel'], Boolean);
  testComponent.propsAreType(CvToggle, ['modelValue'], [Array, Boolean]);

  testComponent.propsHaveDefault(CvToggle, ['formItem']);
  testComponent.propsHaveDefaultOfUndefined(CvToggle, ['modelValue', 'checked']);

  // ***************
  // SNAPSHOT TESTS
  // ***************

  it('should render formItem: default', async () => {
    const propsData = { id: 'test-1', value: 'check-1' };
    const wrapper = await shallow(CvToggle, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render when formItem: true && disabled', async () => {
    const propsData = { id: 'test-1', formItem: true, value: 'check-1', disabled: true };
    const wrapper = await shallow(CvToggle, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render when formItem: false && small && hideLabel', async () => {
    const propsData = { id: 'test-1', formItem: false, value: 'check-1', small: true, hideLabel: true };
    const wrapper = await shallow(CvToggle, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
  it('computed isChecked should be false when checked prop is false', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: false };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  it('computed isChecked should be true when checked prop is true', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: true };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  // Changes to checked property affect the computed value isChecked (wrapper.vm.isChecked)
  it('computed isChecked should be true when checked prop was changed to true', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: false };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
    await setProps(wrapper, { checked: true });
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  // Changes to checked property affect the computed value isChecked (wrapper.vm.isChecked)
  it('computed isChecked should be false when checked prop was changed to false', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: true };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
    await setProps(wrapper, { checked: false });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  // Using modelValue property overrides checked.
  it('computed isChecked should be true when checked prop is false and modelValue is true', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: true, checked: false };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  // Using modelValue property overrides checked.
  it('computed isChecked should be false when checked prop is true and modelValue is false', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: false, checked: true };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  // Using an array with modelValue functions as expected.
  it('computed isChecked should be false when modelValue array is empty', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: [] };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  // Using an array with modelValue functions as expected.
  it('computed isChecked should be false when modelValue array does not contain the value', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: ['check-2', 'check-3'] };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  // Using an array with modelValue functions as expected.
  it('computed isChecked should be true when modelValue array contains the value', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: ['check-2', 'check-3', 'check-1'] };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  it('input should have toggle type', async () => {
    const propsData = { formItem: true, value: 'check-1' };
    const wrapper = await shallow(CvToggle, { propsData });
    expect(wrapper.find('input').element.type).toBe('checkbox');
  });

  it('should emit change event on click', async () => {
    const propsData = { formItem: true, value: 'check-1' };
    const wrapper = await shallow(CvToggle, { propsData });
    await trigger(wrapper.find('input'), 'click');
    expect(wrapper.emitted().change).toBeTruthy();
  });

  it('should emit modelEvent event on click', async () => {
    const propsData = { formItem: true, value: 'check-1' };
    const wrapper = await shallow(CvToggle, { propsData });
    await trigger(wrapper.find('input'), 'click');
    expect(wrapper.emitted().modelEvent).toBeTruthy();
  });

  // Clicking updates the value isChecked (wrapper.vm.isChecked)
  it('clicking updates the value isChecked from false to true', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: false };
    const wrapper = await shallow(CvToggle, { propsData });
    await trigger(wrapper.find('input'), 'click');
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  // Clicking updates the value isChecked (wrapper.vm.isChecked)
  it('clicking updates the value isChecked from true to false', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: true };
    const wrapper = await shallow(CvToggle, { propsData });
    await trigger(wrapper.find('input'), 'click');
    expect(wrapper.vm.isChecked).toEqual(false);
  });
});

// describe('CvToggleSkeleton', () => {
//   // ***************
//   // PROP CHECKS
//   // ***************

//   // ***************
//   // SNAPSHOT TESTS
//   // ***************
//   describe('Renders as expected', () => {
//     const wrapper = await shallow(CvToggleSkeleton);

//     expect(wrapper.html()).toMatchSnapshot();
//   });

//   // ***************
//   // FUNCTIONAL TESTS
//   // ***************
// });
