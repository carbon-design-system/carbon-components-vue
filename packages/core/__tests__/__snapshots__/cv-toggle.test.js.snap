// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvToggle should render formItem: default 1`] = `
<div class="bx--form-item"><input type="checkbox" id="test-1" aria-checked="false" class="bx--toggle-input" value="check-1"> <label for="test-1" class="bx--toggle-input__label">

    <span class="bx--toggle__switch"><svg width="6px" height="5px" viewBox="0 0 6 5" class="bx--toggle__check"><path d="M2.2 2.7L5 0 6 1 2.2 5 0 2.7 1 1.5z"></path></svg> <span aria-hidden="true" class="bx--toggle__text--off">Off</span> <span aria-hidden="true" class="bx--toggle__text--on">On</span></span></label></div>
`;

exports[`CvToggle should render when formItem: false && small && hideLabel 1`] = `
<div class=""><input type="checkbox" id="test-1" aria-checked="false" class="bx--toggle-input bx--toggle-input--small" value="check-1"> <label for="test-1" class="bx--toggle-input__label">

    <span class="bx--toggle__switch"><svg width="6px" height="5px" viewBox="0 0 6 5" class="bx--toggle__check"><path d="M2.2 2.7L5 0 6 1 2.2 5 0 2.7 1 1.5z"></path></svg> <span aria-hidden="true" class="bx--toggle__text--off">Off</span> <span aria-hidden="true" class="bx--toggle__text--on">On</span></span></label></div>
`;

exports[`CvToggle should render when formItem: true && disabled 1`] = `
<div class="bx--form-item"><input type="checkbox" id="test-1" aria-checked="false" disabled="disabled" class="bx--toggle-input" value="check-1"> <label for="test-1" class="bx--toggle-input__label">

    <span class="bx--toggle__switch"><svg width="6px" height="5px" viewBox="0 0 6 5" class="bx--toggle__check"><path d="M2.2 2.7L5 0 6 1 2.2 5 0 2.7 1 1.5z"></path></svg> <span aria-hidden="true" class="bx--toggle__text--off">Off</span> <span aria-hidden="true" class="bx--toggle__text--on">On</span></span></label></div>
`;
